### DISCUSSION TITLE
XUnit: Test discovery is taking more than 1 hour

### DISCUSSION THREAD

**Discussion by pawanbrn** [xunit/xunit#2548] [Question] [answered]
We are using XUnit for API tests and currently we have more than 3k tests but it takes 63 minutes of time to actually discover it and then only its ready to execute. 

XUnit: 2.4.1
XUnit.Runner.VisualStudio: 2.4.5
Visual Studio 2019 (Version: 16.11.15)
Visual Studio 2022 (Version: 17.2.3)

I tried in both VS but same results.

========== Test discovery finished: 3515 Tests found in 63 min ==========

**Comment by bradwilson** [comment, marked as answer]
It's hard to guess what might be going on here without a repro project.

Discovery loads the test DLL and then uses reflection to find all public types, then looking for test methods within them. The closest analogue we have is xunit.v3.assert.tests, with ~ 1000 tests, which discovers and runs them all in less than a quarter second on my development machine:

```
xUnit.net v3 In-Process Runner v0.1.1-pre.84+a9a5206405 (64-bit .NET Core 3.1.27)
  Discovering: xunit.v3.assert.tests (method display = ClassAndMethod, method display options = None)
  Discovered:  xunit.v3.assert.tests (found 852 test cases)
  Starting:    xunit.v3.assert.tests (parallel test collections = on, max threads = unlimited, explicit = off)
  Finished:    xunit.v3.assert.tests
=== TEST EXECUTION SUMMARY ===
   xunit.v3.assert.tests  Total: 908, Errors: 0, Failed: 0, Skipped: 0, Not Run: 0, Time: 0.142s
```

There are many places where your code could be running here, causing the slowdown. The most obvious would be if you were doing lengthy operations (perhaps ones that are using a network? and maybe timing out?) when returning theory data. A quick check to see if that's the case would involve temporarily [disabling theory pre-enumeration during discovery by setting `preEnumerateTheories` to `false`](https://xunit.net/docs/configuration-files#preEnumerateTheories).

**Reply by bradwilson** [reply]
Please open an issue with a repro project. Thanks!
